R-package/src/xgboost_R.cc

```diff
@@ -562,33 +705,104 @@ XGB_DLL SEXP XGBoosterCreate_R(SEXP dmats) {
-  BoosterHandle handle;
+  char const **feature_names;
+  bst_ulong len_feature_names = 0;
+  CHECK_CALL(XGDMatrixGetStrFeatureInfo(R_ExternalPtrAddr(dmat),
+                                        "feature_name",
+                                        &len_feature_names,
+                                        &feature_names));
+  if (len_feature_names) {
+    CHECK_CALL(XGBoosterSetStrFeatureInfo(R_ExternalPtrAddr(booster),
+                                          "feature_name",
+                                          feature_names,
+                                          len_feature_names));
+  }
+
+  char const **feature_types;
+  bst_ulong len_feature_types = 0;
+  CHECK_CALL(XGDMatrixGetStrFeatureInfo(R_ExternalPtrAddr(dmat),
+                                        "feature_type",
+                                        &len_feature_types,
+                                        &feature_types));
+  if (len_feature_types) {
+    CHECK_CALL(XGBoosterSetStrFeatureInfo(R_ExternalPtrAddr(booster),
+                                          "feature_type",
+                                          feature_types,
+                                          len_feature_types));
+  }
+  R_API_END();
+  return R_NilValue;
+}
+
+XGB_DLL SEXP XGBoosterSetStrFeatureInfo_R(SEXP handle, SEXP field, SEXP features) {
+  R_API_BEGIN();
+  SEXP field_char = Rf_protect(Rf_asChar(field));
+  bst_ulong len_features = Rf_xlength(features);
 
   int res_code;
   {
-    std::vector<void*> dvec(len);
-    for (R_xlen_t i = 0; i < len; ++i) {
-      dvec[i] = R_ExternalPtrAddr(VECTOR_ELT(dmats, i));
+    std::vector<const char*> str_arr(len_features);
+    for (bst_ulong idx = 0; idx < len_features; idx++) {
+      str_arr[idx] = CHAR(STRING_ELT(features, idx));
     }
-    res_code = XGBoosterCreate(BeginPtr(dvec), dvec.size(), &handle);
+    res_code = XGBoosterSetStrFeatureInfo(R_ExternalPtrAddr(handle),
+                                          CHAR(field_char),
+                                          str_arr.data(),
+                                          len_features);
   }
   CHECK_CALL(res_code);
-  R_SetExternalPtrAddr(R_handle, handle);
-  R_RegisterCFinalizerEx(R_handle, _BoosterFinalizer, TRUE);
+  Rf_unprotect(1);
   R_API_END();
   return R_NilValue;
 }
 
+XGB_DLL SEXP XGBoosterGetStrFeatureInfo_R(SEXP handle, SEXP field) {
+  R_API_BEGIN();
+  bst_ulong len;
+  const char **out_features;
+  SEXP field_char = Rf_protect(Rf_asChar(field));
+  CHECK_CALL(XGBoosterGetStrFeatureInfo(R_ExternalPtrAddr(handle),
+                                        CHAR(field_char), &len, &out_features));
+  SEXP out = Rf_protect(Rf_allocVector(STRSXP, len));
+  for (bst_ulong idx = 0; idx < len; idx++) {
+    SET_STRING_ELT(out, idx, Rf_mkChar(out_features[idx]));
+  }
+  Rf_unprotect(2);
+  return out;
+  R_API_END();
+  return R_NilValue; /* <- should not be reached */
+}
+
+XGB_DLL SEXP XGBoosterBoostedRounds_R(SEXP handle) {
```

Comment:
Can we extract some of the unrelated changes into a smaller PR and merge them first? Makes reviews easier. ;-)

Reply:
It'd be quite hard to do so, because this piece of information (which replaces both of the older 'niter') is used internally in combination with other attributes from C and R.

It's also not possible to make this PR without these sort of changes, because these attributes are now set and accessed differently, and the PR by default leaves the booster clean of custom C attributes, which requires this functionality.